File: pe-lwjgl-binding/src/main/java/org/saar/lwjgl/opengl/fbos/IFbo.java
package org.saar.lwjgl.opengl.fbos;

import org.saar.lwjgl.opengl.utils.GlBuffer;

public interface IFbo extends ReadOnlyFbo, ReadableFbo, DrawableFbo {

    /**
     * Blit the fbo into the bound read fbo
     */
    void blitFramebuffer(int x1, int y1, int w1, int h1, int x2, int y2, int w2,
                         int h2, FboBlitFilter filter, GlBuffer... buffers);

    /**
     * Sets the size of the fbo
     * *NOTE* you must resize the attachments too
     *
     * @param width  the width
     * @param height the height
     */
    void resize(int width, int height);

    /**
     * Delete the fbo
     */
    void delete();

}

File: pe-lwjgl-binding/src/main/java/org/saar/lwjgl/opengl/fbo/IFbo.kt
package org.saar.lwjgl.opengl.fbo

import org.saar.lwjgl.opengl.fbo.attachment.IAttachment
import org.saar.lwjgl.opengl.fbo.attachment.index.AttachmentIndex
import org.saar.lwjgl.opengl.fbo.rendertarget.DrawRenderTarget
import org.saar.lwjgl.opengl.fbo.rendertarget.ReadRenderTarget

interface IFbo : ReadOnlyFbo {
    /**
     * Adds an attachment to the fbo
     * 
     * @param index      the index of the attachment
     * @param attachment the attachment to add
     */
    fun addAttachment(index: AttachmentIndex, attachment: IAttachment)

    /**
     * Sets the draw target of the fbo
     * 
     * @param target the draw target
     */
    fun setDrawTarget(target: DrawRenderTarget)

    /**
     * Sets the read source of the fbo
     * 
     * @param target the read source
     */
    fun setReadTarget(target: ReadRenderTarget)

    /**
     * Delete the fbo
     */
    fun delete()
}
